### ISSUE TITLE
vendor issues - rfi (request for ideas/improvments)

### ISSUE THREAD

**Issue by ldemailly** [istio/istio#4746]
Despite https://github.com/istio/istio/wiki/Vendor-FAQ#how-do-i-add--change-a-dependency

The vendor changes are more error prone and painful than they should be

There are a couple of distinct issues

- The fundamental one - which wouldn't go away in any setup I can think of, is vendor changes are intrinsically serialized / not mergeable (though they could be mostly mergeable - or we could automate the merging, and recalculation of go dep's input sha)

- Pruning is a good thing but also means anything not in use is deleted, before it can be used.

- Istio API changes should be automatically tracked (and fortio's too while at it)

- Contributors need to make a branch in vendor-istio which they can't do by default (only core contributors can) - I have been adding people one by one but that's not scaleable

I'm quite tempted to throw the towel at the coming end of my 2months warranty period and let people check in directly into istio/istio

cc @jwendell @ZackButcher @sebastienvas @hklai

**Comment by hklai** [comment]
Having someone babysit this is certainly not sustainable. However, if we let people check in dependencies in istio/istio, do we anticipate a different set of problems? The main concern I remember was code ownership, but that was more on the philosophical side. Are there more practical problems that we may face?

**Comment by ldemailly** [comment]
The practical problem would be conflicts between PRs because of vendor changes in multiple PRs

And vendor creep

And some amount of starting over on education (updating the faq etc...)

(in addition to inability to undo mistakes like when 20k files are added which happened early on on vendor because someone used an older dep which didn't understand the new-ish pruning option, I erased that history there, etc...
and the philosophy of a large chunk of foreign files checked in in the main repo)

**Comment by sebastienvas** [comment]
Could we have a check that make sure that vendors are updated correctly ? and make it required ?

**Comment by nmittler** [comment]
FWIW, I've been trying to add my first dependency and the experience is pretty rough.

The main problem that I ran into is that I tried adding my dependency while another developer was in the middle of adding theirs (and therefore had the lock). When I tried adding my dep, it removed theirs ... even if I ran `dep ensure --add <my dep>`.

Few, if any other Istio developers seem to really understand this machinery, so I was mostly relying on the documentation in the [FAQ](https://github.com/istio/istio/wiki/Vendor-FAQ), which is not sufficient and a bit hard to follow. There doesn't seem to be anything that clearly documents the general workflow for adding a dependency.  What I've learned in talking with folks is that there are a total of 3 PRs: 1 to change to the vendor repo, 1 to import the new vendor repo into istio/istio, and one to add your changes that use the new dependency. This is not a sustainable pattern.

I think, in the end, there are only 2 real solutions for fixing our developer workflow:

1. Get rid of vendor from the repo entirely and have the build regenerate it every time. The main concern here is reliable builds. So, maybe not a viable option.
2. Check in the entire vendor directory into istio/istio and get rid of the submodule entirely.

In all cases, we'll have to be disciplined with adding dependencies (i.e. reviews that involve adding deps have to have more scrutiny) ... we can't escape that in any case. Regarding [@hklai's comment on code ownership](https://github.com/istio/istio/issues/4746#issuecomment-379136802), we're all owners ... and it's the same regardless of whether the vendor directory is in istio/istio or elsewhere.

AFAICT there is no advantage to having vendor outside of istio/istio, but the separation adds a lot of complexity and difficulty to the developer workflow.  I propose that we just move it into istio/istio and call it a day.

@sebastienvas @hklai @ldemailly @ZackButcher  thoughts?

**Comment by knrc** [comment]
@nmittler +1 on moving it into istio/istio, when we originally proposed the inclusion of vendor this was what we were intending especially with the general problems we demonstrated with submodules.

The compromise we reached in one of the meetings was to try the submodule approach and see how well/badly it worked and then revisit after a couple of months, I'm pleased to see this is now happening.

**Comment by hklai** [comment]
"We don't own or author the code we submit to vendor" - that was the ownership problem I was referring to, and that was one of the reasons why some people wanted a separate repo instead.

Now that we experimented, perhaps the convenience of submitting to istio/istio outweighs this problem.

@knrc @ldemailly @nmittler could you attend the test/release group meeting next Monday 11 am to make a decision?

**Comment by ldemailly** [comment]
Note that if we do proceed with the move of istio API inside istio, a large chunk of the downsides goes away

Also once done once, the process is quite repeatable (kuat, Zack, ... Can do it fine , check the commit history in vendor-istio to find who to talk to)

Lastly it's 2 PRs not 3, you have to use the dependencies you introduce or prune will drop them

**Comment by kyessenov** [comment]
I agree with what Laurent has said. We should not be optimizing for this
case, since we don't really want proliferation of dependencies. Some cases
arise due to "pruning" of large dependencies, like k8s; or dep trying to
pro-actively remove package usages. This can be dealt with separately, and
is going to be an issue even if we commit vendor.

The sequencing of vendor updates is also not solved by moving vendor into
istio. We still have merge conflicts and the process is sequential no
matter how you look at it.

**Comment by ZackButcher** [comment]
Moving vendor into istio/istio solves the problem because we can have a single PR introducing new code to Istio with a new dependency, rather than a PR into vendor-istio with a dependency followed by one into istio/istio using that dependency (the PR into vendor-istio effectively takes a global lock on vendor-istio until the istio/istio PR is merged).

Even though yes I have successfully added a dependency, it was not an easy process. I was around to give @nmittler tips and it was still a painful process for him. I was initially opposed to moving vendor into istio/istio proper but I'm more in favor of it now.

Making developer workflow hard is never a solution to any problem in the project. If we're worried that making it easier to add dependencies will cause a proliferation of dependencies, then lock down OWNERS of the vendor dir to a few people who vet dependencies are they're added. At a minimum we should be able to automate more of the process via make, I think.

**Comment by nmittler** [comment]
Agree with @ZackButcher. I still haven't seen a strong argument for keeping the vendors repository separate. Unfortunately, the process is *very* repeatable ... it's a gigantic pain every time you do it.  Keep in mind, I wasn't doing anything "wrong", the problem is the process itself. We need to optimize for making the developer workflow simpler for everyone.  If we want to make modifications to vendors harder, OWNERS should help.  In my mind, a 1-PR solution a hard requirement.

**Comment by ldemailly** [comment]
interesting use case: https://github.com/istio/istio/pull/4877
had committed 30k files and 9M lines changed in vendor which thankfully I was able to erase before it permanently damages the repo. That isn't possible if we don't have a submodule

(it also unfortunately shows that the process is "hard"...)

https://github.com/istio/vendor-istio/pull/55

**Comment by ldemailly** [comment]
we can save 1 step using merge commit instead of squash and merge (the sha in the 'branch' still exists)

**Comment by ldemailly** [comment]
'Fixed' by #5014

Thanks for picking up the burden on this on nate

**Comment by ldemailly** [comment]
follow up in #5023
